File: Heap/Python/Last Stone Weight.py
# Problem link: https://leetcode.com/problems/last-stone-weight/description/
# Time: O(n * log n)
# Space: O(n)
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        maxStones = [-s for s in stones]
        heapq.heapify(maxStones)

        while len(maxStones) > 1:
            heavyOne = -heapq.heappop(maxStones)
            heavyTwo = -heapq.heappop(maxStones)

            if heavyOne != heavyTwo:
                heapq.heappush(maxStones, -(heavyOne - heavyTwo))

        return -maxStones[0] if maxStones else 0

File: Heap/Java/Last Stone Weight.java
// Problem link: https://leetcode.com/problems/last-stone-weight/description/
// Time: O(n log n)
// Space: O(n)
class Solution {
    public int lastStoneWeight(int[] stones) {
        PriorityQueue<Integer> maxStones = new PriorityQueue<>();
        int heavyOne, heavyTwo;

        for(int s : stones) {
            maxStones.offer(-s);
        }

        while(maxStones.size() > 1) {
            heavyOne = -maxStones.poll();
            heavyTwo = -maxStones.poll();

            if(heavyOne != heavyTwo) {
                maxStones.offer(-(heavyOne - heavyTwo));
            }
        }

        return maxStones.size() > 0 ? -maxStones.poll() : 0;
    }
}
